sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -242,8 +239,10 @@ public CompositeBehavior enterCompositeTransform(Node node) {
           @Override
           public void leaveCompositeTransform(Node node) {
             if (node.isRootNode()) {
-              checkState(
```

Comment:
why this is no longer the case? To my understanding we should still have some check. If this change adds an exception we can check the either matched is empty or contains the known transform that leaves out.

Reply:
This is because if autosharding is enabled in the transform, we return the replacement transform with autosharding enabled, and DO NOT set a fixed number of shards (so the transform is the same as before) here: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java   Previously, we’d modify the transform, which is why it would not be the same transform. 
   overridesBuilder.add(
            PTransformOverride.of(
                PTransformMatchers.writeWithRunnerDeterminedSharding(),
                new StreamingShardedWriteFactory(options)));


I'm not sure that I understand why we have this constraint, after the translation, why must all nodes be different from the original graph?

Reply:
This was a very old check, presented 7 years ago:

https://github.com/apache/beam/commit/85af898132ec5c528a96d5c213c08cee91fa6538

Not sure if anyone still has the context.

I still not quite understand what exactly happened, as I am not quite familiar with this codepath either. In any case, can we preserve the original check like
```
if (node.isRootNode() && !matched.isEmpty()) {
  LOG.info("Found nodes that matched overrides. Matches: %s", matched. The match usually should be empty unless there are runner specific replacement transform.");
}
```
if this phrasing is correct.

Reply:
Thanks for the context, the suggested log you've provided is correct.  Thanks!